lib/universal/mapper.js

```diff
@@ -0,0 +1,785 @@
+  if (screen.height && screen.width) {
+    form.sr = fmt('%sx%s', screen.width, screen.height);
+  }
+
+  // locale
+  if (locale) form.ul = locale;
+
+  // app
+  if (app.name) form.an = app.name;
+  if (app.version) form.av = app.version;
+  if (app.appId) form.aid = app.appId;
+  if (app.appInstallerId) form.aiid = app.appInstallerId;
+
+  if (settings.sendUserId && facade.userId()) form.uid = facade.userId();
+  if (facade.userAgent()) form.ua = facade.userAgent();
+  if (facade.ip()) form.uip = facade.ip();
+
+  return form;
+}
+
+/**
+ * Adds hostname and path
+ *
+ * @api private
+ * @param {Object} facade
+ * @param {Object} form
+ * @return {Object}
+ */
+
+function createPageDataForm(facade, form){
+  var url = facade.proxy('properties.url') || facade.proxy('context.page.url');
+  if (!is.string(url)) return form;
+
+  var parsed = parse(url || '');
+  if (parsed.hostname) form.dh = parsed.hostname;
+  if (parsed.path) form.dp = parsed.path;
+
+  return form;
+}
+
+/**
+ * Map google's custom dimensions & metrics with `obj`.
+ *
+ * Example:
+ *
+ *      metrics({ revenue: 1.9 }, { { metrics : { revenue: 'metric8' } });
+ *      // => { metric8: 1.9 }
+ *
+ *      metrics({ revenue: 1.9 }, {});
+ *      // => {}
+ *
+ * @api private
+ * @param {Object} obj
+ * @param {Object} data
+ * @return {Object|null}
+ */
+
+function metrics(obj, data){
+  var dimensions = data.dimensions || {};
+  var metrics = data.metrics || {};
+  var names = Object.keys(metrics).concat(Object.keys(dimensions));
+  var ret = {};
+
+  for (var i = 0; i < names.length; ++i) {
+    var name = names[i];
+    var key = shorten(metrics[name] || dimensions[name]);
+    var value = lookup(obj, name) || obj[name];
+    if (value == null) continue;
+    if (is.boolean(value)) value = value.toString();
+    if (key) ret[key] = value;
+  }
+
+  return ret;
+}
+
+/**
+ * Shorten `metric\d+` or `dimension\d+` to `cm\d+`, `cd\d+`.
+ *
+ * Example:
+ *
+ *    shorten('metric99'); // => cm99
+ *    shorten('dimension57'); // => cd57
+ *
+ * @api private
+ * @param {string} name
+ * @return {string}
+ */
+
+function shorten(name){
+  var match = name.match(/^metric(\d+)$/);
+  if (match) return 'cm' + match[1];
+  match = name.match(/^dimension(\d+)$/);
+  if (match) return 'cd' + match[1];
+}
+
+/**
+ * Determine whether the request is being made directly on behalf of a
+ * mobile device (iOS, android, Xamarin).
+ *
+ * @api private
+ * @param {Object|String} library
+ * @return {Boolean}
+ */
+
+function isMobile(library){
+  if (!library) return false;
+  return contains('ios') || contains('android') || contains('analytics.xamarin');
+
+  function contains(str) {
+    return name().toLowerCase().indexOf(str) !== -1;
+  }
+
+  function name() {
+    return typeof library === 'string' ? library : library.name;
+  }
+}
+
+/**
+ * Return a copy of an object, less an  `undefined` values.
+ *
+ * @param {Object} obj
+ * @return {Object}
+ */
+
+function compactObject(obj) {
```

Comment:
On that TODO note above about filtering out any undefined properties—this would be a useful util for that